spec/integration/http/client_spec.rb

```diff
@@ -0,0 +1,140 @@
+    # make sure we don't take too long
+    Puppet[:http_connect_timeout] = '5s'
+  end
+
+  let(:hostname) { '127.0.0.1' }
+  let(:wrong_hostname) { 'localhost' }
+  let(:server) { PuppetSpec::HTTPSServer.new }
+  let(:client) { Puppet::HTTP::Client.new }
+  let(:ssl_provider) { Puppet::SSL::SSLProvider.new }
+  let(:root_context) { ssl_provider.create_root_context(cacerts: [server.ca_cert], crls: [server.ca_crl]) }
+
+  context "when verifying an HTTPS server" do
+    it "connects over SSL" do
+      server.start_server do |port|
+        res = client.get(URI("https://127.0.0.1:#{port}"), ssl_context: root_context)
+        expect(res).to be_success
+      end
+    end
+
+    it "raises if the server's cert doesn't match the hostname we connected to" do
+      server.start_server do |port|
+        expect {
+          client.get(URI("https://#{wrong_hostname}:#{port}"), ssl_context: root_context)
+        }.to raise_error { |err|
+          expect(err).to be_instance_of(Puppet::HTTP::ConnectionError)
+          expect(err.message).to match(/Server hostname '#{wrong_hostname}' did not match server certificate; expected one of (.+)/)
+
+          md = err.message.match(/expected one of (.+)/)
+          expect(md[1].split(', ')).to contain_exactly('127.0.0.1', 'DNS:127.0.0.1', 'DNS:127.0.0.2')
+        }
+      end
+    end
+
+    it "raises if the server's CA is unknown" do
+      wrong_ca = cert_fixture('netlock-arany-utf8.pem')
+      alt_context = ssl_provider.create_root_context(cacerts: [wrong_ca], revocation: false)
+
+      server.start_server do |port|
+        expect {
+          client.get(URI("https://127.0.0.1:#{port}"), ssl_context: alt_context)
+        }.to raise_error(Puppet::HTTP::ConnectionError,
+                         %r{certificate verify failed.* .self signed certificate in certificate chain for CN=Test CA.})
+      end
+    end
+  end
+
+  context "with client certs" do
+    let(:ctx_proc) {
+      -> ctx {
+        # configures the server to require the client to present a client cert
+        ctx.verify_mode = OpenSSL::SSL::VERIFY_PEER | OpenSSL::SSL::VERIFY_FAIL_IF_NO_PEER_CERT
+      }
+    }
+
+    it "raises if the server requires a client cert, but we don't have one" do
+      server.start_server(ctx_proc: ctx_proc) do |port|
+        if OpenSSL::OPENSSL_VERSION_NUMBER >= 0x10101000
+          # TLS 1.3 uses post-handshake authentication, so the client cert is sent
+          # after the handshake. For this test, we don't send a client cert, so
+          # the server aborts the connection as soon as we send application data.
+          # This results in a generic HTTPError instead of a ConnectionError. This
+          # test assumes TLS 1.3 is always selected when using openssl 1.1 and up
+          expect {
+            client.get(URI("https://127.0.0.1:#{port}"), ssl_context: root_context)
+          }.to raise_error(Puppet::HTTP::HTTPError, %r{An existing connection was forcibly closed by the remote host})
```

Comment:
I'm still having issues with this test. I'm getting a timeout error rather than force close from the remote host

```
...
From: /Users/melissa.stone/workdir/puppet/spec/integration/http/client_spec.rb @ line 73 :

    68:       }
    69:     }
    70:
    71:     it "raises if the server requires a client cert, but we don't have one" do
    72:       server.start_server(ctx_proc: ctx_proc) do |port|
 => 73:         require 'pry'; binding.pry
    74:         if OpenSSL::OPENSSL_VERSION_NUMBER >= 0x10101000
    75:           # TLS 1.3 uses post-handshake authentication, so the client cert is sent
    76:           # after the handshake. For this test, we don't send a client cert, so
    77:           # the server aborts the connection as soon as we send application data.
    78:           # This results in a generic HTTPError instead of a ConnectionError. This

[1] pry(#<RSpec::ExampleGroups::PuppetHTTPClient::WithClientCerts>)> OpenSSL::OPENSSL_VERSION_NUMBER
=> 269488207
[2] pry(#<RSpec::ExampleGroups::PuppetHTTPClient::WithClientCerts>)> continue
F....

Failures:

  1) Puppet::HTTP::Client with client certs raises if the server requires a client cert, but we don't have one
     Failure/Error:
       expect {
         client.get(URI("https://127.0.0.1:#{port}"), ssl_context: root_context)
       }.to raise_error(Puppet::HTTP::HTTPError, %r{An existing connection was forcibly closed by the remote host})

       expected Puppet::HTTP::HTTPError with message matching /An existing connection was forcibly closed by the remote host/, got #<Puppet::HTTP::HTTPError: Request to https://127.0.0.1:56173 timed out after 5.006 seconds> with backtrace:
         # ./lib/puppet/http/client.rb:127:in `rescue in handle_post_connect'
         # ./lib/puppet/http/client.rb:119:in `handle_post_connect'
         # ./lib/puppet/http/client.rb:25:in `block in connect'
         # ./lib/puppet/network/http/pool.rb:31:in `with_connection'
         # ./lib/puppet/http/client.rb:23:in `connect'
         # ./lib/puppet/http/client.rb:80:in `block in execute_streaming'
         # ./lib/puppet/http/client.rb:79:in `loop'
         # ./lib/puppet/http/client.rb:79:in `execute_streaming'
         # ./lib/puppet/http/client.rb:43:in `get'
         # ./spec/integration/http/client_spec.rb:81:in `block (5 levels) in <top (required)>'
         # ./spec/integration/http/client_spec.rb:80:in `block (4 levels) in <top (required)>'
         # ./spec/lib/puppet_spec/https.rb:70:in `block in start_server'
         # ./spec/lib/puppet_spec/https.rb:30:in `pipe'
         # ./spec/lib/puppet_spec/https.rb:30:in `start_server'
         # ./spec/integration/http/client_spec.rb:72:in `block (3 levels) in <top (required)>'
         # ./spec/spec_helper.rb:176:in `block (2 levels) in <top (required)>'
     # ./spec/integration/http/client_spec.rb:80:in `block (4 levels) in <top (required)>'
     # ./spec/lib/puppet_spec/https.rb:70:in `block in start_server'
     # ./spec/lib/puppet_spec/https.rb:30:in `pipe'
     # ./spec/lib/puppet_spec/https.rb:30:in `start_server'
     # ./spec/integration/http/client_spec.rb:72:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:176:in `block (2 levels) in <top (required)>'

Finished in 21.14 seconds (files took 0.99685 seconds to load)
8 examples, 1 failure

Failed examples:

rspec ./spec/integration/http/client_spec.rb:71 # Puppet::HTTP::Client with client certs raises if the server requires a client cert, but we don't have one
                                                                                                                       /22.5s

in puppet on  create_system_context_10144 [!] via 💎 v2.4.6
➜ bundle exec openssl version
LibreSSL 2.6.5
```

Reply:
yeah something is not right with your libressl. I'd recommend upgrading your ruby-build and reinstalling a new version of ruby:

```
$ brew upgrade ruby-build
$ rbenv install 2.4.9
```

Or even 2.5.7:

```
$ rbenv install 2.5.7
Downloading openssl-1.1.1d.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/1e3a91bc1f9dfce01af26026f856e064eab4c8ee0a8f457b5ae30b40b8b711f2
Installing openssl-1.1.1d...
...
$ gem install --no-document bundler
...
$ cd puppet
$ bundle install --path .bundle
```

Newer versions of ruby-build select the appropriate openssl version to install for that particular ruby version.

Reply:
I'm getting a different error when using ruby 2.5.7 + openssl 1.1.1... let me take a look:

```
       expected Puppet::HTTP::HTTPError with message matching /An existing connection was forcibly closed by the remote host/, got #<Puppet::HTTP::HTTPError: Request to https://127.0.0.1:55269 failed after 0.0 seconds: SSL_read: tlsv13 alert certificate required> with backtrace:
         # ./lib/puppet/http/client.rb:129:in `rescue in handle_post_connect'
         # ./lib/puppet/http/client.rb:119:in `handle_post_connect'
         # ./lib/puppet/http/client.rb:25:in `block in connect'
```

Reply:
Filed as PUP-10154

Reply:
are we going to pull out the integration test until pup-10154, or leave it as is?

Reply:
Going to leave the integration tests except for the client cert test, which I'll handle later.

Reply:
Also remembered that the `PuppetSpec::HTTPSServer` class was added awhile ago, and is used in `spec/integration/network/http_pool_spec.rb`. Do those tests fail too, or is something in this PR?

Reply:
yup, that's good

```
in puppet on  create_system_context_10144 via 💎 v2.6.3
➜ be rspec spec/integration/network
Run options: exclude {:broken=>true, :benchmark=>true}
...........**......................................................

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) Puppet::Network::AuthConfig allow should support hostname ending with '.'
     # bug #7589
     Failure/Error: expect(@auth).to auth_allow(request(:node => 'host.domain.com.'))
       expected host.domain.com./10.1.1.1 to be allowed
     # ./spec/integration/network/authconfig_spec.rb:130:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:176:in `block (2 levels) in <top (required)>'

  2) Puppet::Network::AuthConfig allow should support hostname ending with '.' and backreferences
     # bug #7589
     Failure/Error: expect(@auth).to auth_allow(request(:node => 'host.domain.com.'))
       expected host.domain.com./10.1.1.1 to be allowed
     # ./spec/integration/network/authconfig_spec.rb:137:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:176:in `block (2 levels) in <top (required)>'

Finished in 1.34 seconds (files took 0.96862 seconds to load)
67 examples, 0 failures, 2 pending
```